api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,203 @@
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.mutation.v3alpha";
+option java_outer_classname = "MutationProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: External Mutation Service]
+
+// A service that can mutate HTTP requests and responses.
+// The overall mutation protocol works like this:
+//
+// 1. Envoy sends to the service information about the HTTP request.
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server a chunk of the message body.
+//    and the server sends back a MutationResponse.
+// 4. If so requested, Envoy sends the server the HTTP trailers,
+//    and the server sends back a MutationResponse.
+// 5. At this point, request processing is done, and we pick up again
+//    at step 1 when Envoy receives a response from the upstream server.
+// 6. At any point above, if the server closes the gRPC stream cleanly,
+//    then Envoy proceeds without consulting the server.
+// 7. At any point above, if the server closes the gRPC stream with an error,
+//    then Envoy returns a 500 error to the client, unless the filter
+//    was configured to ignore errors.
+//
+// In other words, the process is a request/response conversation, but
+// using a gRPC stream to make it easier for the client and server to
+// maintain state. The exception is the "SEND_BODY_ASYNC" command,
+// which requests that the server keep sending body chunks without waiting
+// for a response.
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+message MutationRequest {
+  oneof r {
+    // Information about the HTTP request, sent after the HTTP headers have been
+    // received on the "encoding" path, but before the body is available.
+    // This is the same data sent to an "ext_authz" service.
+    auth.v3.AttributeContext request = 1;
+
+    // Information about the HTTP response, sent after the HTTP headers have been
+    // received on the "decoding" path.
+    Response response = 2;
+
+    // A chunk of the HTTP request body. This is only sent if the server
+    // requested it.
+    Body body = 3;
+
+    // The HTTP trailers, for either the request or response path, depending
+    // on where we are in the filter lifecycle. This is only sent if the
```

Comment:
The order of downstream/upstream trailers isn't fixed, an upstream might send trailers before downstream or vice versa.

Reply:
OK -- I'll have to think about how to prevent the asynchronous nature of all this from getting out of control for the server implementer.

Reply:
I assume it's also possible that encodeHeaders could come before decodeData, if the upstream has returned an error before the downstream has sent the whole body, right?

Reply:
Right.